### ISSUE TITLE
Pre-release extension testing

### ISSUE THREAD

**Issue by mwjames** [SemanticMediaWiki/SemanticCite#1]
@SemanticMediaWiki/testers I'd like to invite people to test this extension before its release. It requires
- SMW@dev (or at least https://github.com/SemanticMediaWiki/SemanticMediaWiki/commit/7c5d853de25441b2a1cb7aca44d381cad0024cac)
- MW 1.24+

## How to install
Just run `composer require mediawiki/semantic-cite:@dev` + `php maintenance/update.php`

## What to test

The general functionality of creating citation resources, display references (and lists), search citations, add individual text formatting and property assignments. See also:
- https://github.com/SemanticMediaWiki/SemanticCite#usage
- https://github.com/SemanticMediaWiki/SemanticCite/blob/master/CONFIGURATION.md#configuration
- Example for a template based formatting can be found at https://github.com/SemanticMediaWiki/SemanticCite/blob/master/tests/phpunit/Integration/Fixtures/scite-04-use-template-definition.json#L10-L26
- Example for property mapping can be found at https://github.com/SemanticMediaWiki/SemanticCite/blob/master/tests/phpunit/Integration/Fixtures/scite-03-use-property-definition.json#L18-L25

**Comment by jaideraf** [comment]
I'll test tomorrow and see what happens. 
It seems a complex and customisable extension. I like it. :+1:

**Comment by mwjames** [comment]
> It seems a complex

I don't hope so because if someone just wanted to start without property mapping and template formatting then a simple  [0] should do it:
```
{{#scite:SomeKindOfReferenceUsedWikiWide
 |type=SomeType
 |citation text=This text will be displayed without further formatting, as-is
}}
```
by adding the reference `[[CiteRef::SomeKindOfReferenceUsedWikiWide]]` to a page.

If you want a template to manage the formatting directly then
```
{{#scite:SomeKindOfReferenceUsedWikiWide
 |type=SomeType
 |authors=AuthorA, AuthorB
 |year=1984
 |publisher=BigBrother
 |template=ATemplateThatWillFormatByTheRuleOfOversightAndOppression
}}
```
 and as last option where you have established types and guidelines, the type can decide [1] on its own which template to use to generate the citation text.

```
{{#scite:SomeKindOfReferenceUsedWikiWide
 |type=ThisTypeHasATemplateAssignedSoTheFormattingHappensOnItsOwn
 |authors=AuthorA, AuthorB
 |year=1984
 |publisher=BigBrother
}}
```

> customisable extension

This is where the property and template mapping comes into play because I didn't want the extension to be bound to any specific citation guideline or property assignment.

Some properties are pre-defined with special output formatting such as PMCID or DOI and others come with extra verification procedures, any other property assignment is managed by [2] since the user should decide what semantics (s)he needs to manage.

The reference list is self-added but if for some reason one wants to place the list within an article or before a template then using `{{#referencelist:}}` should help to solve this as well.

[0] https://github.com/SemanticMediaWiki/SemanticCite/blob/master/CONFIGURATION.md#how-to-create-a-simple-citation-resource

[1] https://github.com/SemanticMediaWiki/SemanticCite/blob/master/CONFIGURATION.md#template-mapping

[2] https://github.com/SemanticMediaWiki/SemanticCite/blob/master/CONFIGURATION.md#property-mapping

**Comment by jaideraf** [comment]
>This is where the property and template mapping comes into play because I didn't want the extension to be bound to any specific citation guideline or property assignment.

Smart choice.

I tested: http://wikincat.org/wiki/User:Jaider/Sandbox
Everything seems OK. 

`__NOREFERENCELIST__` tested: OK
MediaWiki:Sci-template-definition tested: OK
MediaWiki:Sci-property-definition tested: OK
scite.styles.css modified via MediaWiki:Common.css tested: OK
$GLOBALS['scigCitationReferenceCaptionFormat'] tested: OK
$GLOBALS['scigBrowseLinkToCitationResource'] tested: OK
$GLOBALS['scigNumberOfReferenceListColumns'] tested: OK
$GLOBALS['scigReferenceListType'] tested: OK

I only could not make the pages appear using the syntax 

`[[CiteRef::Author1; Author2,  2012|Author1; Author2, 2012, p. 78]]`

Am I doing something wrong?

Suggestion: 
`{{#referencelist:}}` could be extended to

```
{{#referencelist:
 |columnNumber=2
 |listType=ol
 |specialBrowse=no
}}
```

thus, overwriting the default options set in LocalSettings.php (and the default options set in the extension itself).

**Comment by jaideraf** [comment]
@JeroenDeDauw, could you make this extension be available on translatewiki.net?

**Comment by mwjames** [comment]
> I tested: http://wikincat.org/wiki/User:Jaider/Sandbox
Everything seems OK.

Thanks for taking the time to test some of the use cases.

> could you make this extension be available on translatewiki.net?

Done with #2.

> I only could not make the pages appear using the syntax
[[CiteRef::Author1; Author2, 2012|Author1; Author2, 2012, p. 78]]
Am I doing something wrong?

For the example below (for `$GLOBALS['scigCitationReferenceCaptionFormat'] = SCI_CITEREF_KEY;`)
```
Lorem ipsum dolor [[CiteRef::Author1; Author2, 2012]] sit amet consectetuer justo Nam quis lobortis vel. Sapien nulla enim Lorem enim pede lorem nulla justo diam wisi. [[CiteRef::Author1; Author2, 2012|Author1; Author2, 2012, p. 78]]

{{#scite:Author1; Author2, 2012
 |type=sometype
 |citation text=some text
}}

or as

{{#scite:
 |reference=Author1; Author2, 2012
 |type=sometype
 |citation text=some text
}}
```

now `Author1; Author2, 2012, p. 78` should be displayed (fixed with 8f47bae3d4).

> Suggestion: 
{{#referencelist:}} could be extended to

Added with c4a6d969 to accept things like:

```
{{#referencelist:
 |listtype=ol
 |browselinks=true
 |columns=2
 |header=SomeHeader
}}
``

**Comment by jaideraf** [comment]
Now, it's almost perfect!

* When I use `{{#referencelist:|listtype=ul}}`, additionally to the change of ordered list to unordered list, the "a b c d ↑" becomes "4.a 4.b 4.c 4.d ↑".
* When I use `{{#referencelist:|browseLinks=false}}`, additionally to the change of visibility of browse links, it shows the 'Citation key' before 'Citation text'.

Maybe unrelated to this, how can I use the ´@sortkey´ to make the referente list appear in sorted order of 'Citation key' or 'Citation text'? Maybe something like: `{{#referencelist:|sortType=key|sortOrder=asc}}`.

**Comment by JeroenDeDauw** [comment]
@jaideraf see #2

**Comment by mwjames** [comment]
> When I use {{#referencelist:|listtype=ul}}, additionally to the change of ordered list to unordered list, the "a b c d ↑" becomes "4.a 4.b 4.c 4.d ↑".

Fixed with 0eda756fd80.

> When I use {{#referencelist:|browseLinks=false}}, additionally to the change of visibility of browse links, it shows the 'Citation key' before 'Citation text'.

Fixed with 0eda756fd80.

> Maybe unrelated to this, how can I use the ´@sortkey´ to make the referente list appear in sorted order of 'Citation key' or 'Citation text'? Maybe something like: {{#referencelist:|sortType=key|sortOrder=asc}}.

Sorting and the usage of `sortkey` requires some work which can be feature but it won't make it into the 1.0 release.

> Now, it's almost perfect!

Over the weekend I added (@JeroenDeDauw would say more awesomeness): 

- 9658216d5c enables the tooltip for both `SCI_CITEREF_NUM` and `SCI_CITEREF_KEY` by default
- Add `{{#referencelist:|toc=yes}}` to generate a TOC section (for when the TOC is displayed)
- Support for nonbound reference lists #4 (requires https://github.com/SemanticMediaWiki/SemanticMediaWiki/commit/53d3c16396e8fcd2e506bd4f1c441ee9565f61a9)

**Comment by jaideraf** [comment]
Awesome! :+1:

**Comment by JeroenDeDauw** [comment]
@mwjames that list is missing "interoperability with https://github.com/wikimedia/mediawiki-extensions-Awesomeness"

**Comment by jaideraf** [comment]
> We have a mirror on Gerrit. No translation updates there yet though. Perhaps the TWN people need poking?

Just to confirm: Semantic Cite did not appear on translatewiki.net yet. [1]
1. https://translatewiki.net/wiki/Special:Translate?action=translate

**Comment by kghbln** [comment]
I poked yesterday. See this https://github.com/SemanticMediaWiki/SemanticCite/issues/2#issuecomment-119346440

**Comment by mwjames** [comment]
Thanks all involved for testing.

PS: While it was not planned to be part of 1.0, I just merged `Special:FindMetadataById` with 1.0 allowing it to be available for general use.
